libbeat/outputs/kafka/config_test.go

```diff
@@ -97,3 +104,35 @@ func TestConfigInvalid(t *testing.T) {
 		})
 	}
 }
+
+func TestBackoffFunc(t *testing.T) {
+	lbtesting.SeedPRNG(t)
+	err := quick.Check(func(i uint) bool {
+		initBackoff := 1 + (i % 5)               // 1 to 5
+		maxBackoff := initBackoff + 1 + (i % 45) // initBackoff+1 to 50
+		maxRetries := int(1 + (i % 50))          // 1 to 50
+
+		cfg := backoffConfig{
+			Init: time.Duration(initBackoff) * time.Second,
+			Max:  time.Duration(maxBackoff) * time.Second,
+		}
+
+		backoffFn := makeBackoffFunc(cfg)
+
+		for retries := 0; retries < maxRetries; retries++ {
+			backoff := backoffFn(retries, maxRetries)
+
+			if !assert.GreaterOrEqual(t, backoff.Milliseconds(), cfg.Init.Milliseconds()/2) {
+				t.Logf("init: %v, max: %v, retries: %v", cfg.Init, cfg.Max, retries)
+				return false
+			}
+			if !assert.LessOrEqual(t, backoff.Milliseconds(), cfg.Max.Milliseconds()) {
+				t.Logf("init: %v, max: %v, retries: %v", cfg.Init, cfg.Max, retries)
+				return false
+			}
```

Comment:
The range is much too big. It only checks that we are in between init/2 and max, but not the evolution of the values.

Maybe quick check only based testing is some overkill here, while not capturing all cases. How about creating simple tests with `maxRetries=0, maxRetries=1, ...`? Based on this you can prepare an array of the expected exponential backoff values. The limit would be `expected/2 <= current <= expected`.

Reply:
Thanks @urso. I tried to implement the stronger assertions testing the evolution of the backoff values in a8148fc08. Let me know what you think!